File: leetcode-637_average-of-levels-in-binary-tree.js
//求二叉树每一层的平均值
//BFS广度优先遍历
var averageOfLevels = function(root) {
    let list = [], ret = [];
    if (root === null) {
        return ret;
    }
    list.push(root); //list一开始先保存第一层的元素root节点
    while (list.length) {
        let len = list.length; //获取当前层的宽度
        let sum = 0;
        for (let i = 0; i < len; i++) { //遍历当前层时, 将下一层的元素压进list尾部
            let node = list.shift(); //当前层的元素逐个弹出, 当前层遍历完后剩下的元素就全是下一层的
            sum += node.val;
            if (node.left !== null) {
                list.push(node.left);
            }
            if (node.right !== null) {
                list.push(node.right);
            }
        }
        ret.push(sum / len);
    }
    return ret;
};

File: leetcode-637_average-of-levels-in-binary-tree.java
/**
 * 给定一个二叉树，求每一层的平均值
 */
public List<Double> averageOfLevels(TreeNode root) {
    List<Double> ret = new ArrayList<>();
    if (root == null) {
        return ret;
    }
    ArrayList<TreeNode> list = new ArrayList<>();
    list.add(root);
    while (list.size() > 0) {
        int len = list.size();
        double sum = 0;
        for (int i = 0; i < len; i++) { //使用广度优先算法
            TreeNode node = list.remove(0);
            sum += node.val;
            if (node.left != null) {
                list.add(node.left);
            }
            if (node.right != null) {
                list.add(node.right);
            }
        }
        ret.add(sum / len); //求均值
    }
    return ret;
}
